modules/java/generator/src/java/android+JavaCamera2View.java

```diff
@@ -0,0 +1,366 @@
+                    throw new RuntimeException(
+                            "Time out waiting to lock camera opening.");
+                }
+                Log.i(LOGTAG, "Opening camera: " + mCameraID);
+                manager.openCamera(mCameraID, mStateCallback, mBackgroundHandler);
+            }
+        } catch (CameraAccessException e) {
+            Log.e(LOGTAG, "OpenCamera - Camera Access Exception");
+        } catch (IllegalArgumentException e) {
+            Log.e(LOGTAG, "OpenCamera - Illegal Argument Exception");
+        } catch (SecurityException e) {
+            Log.e(LOGTAG, "OpenCamera - Security Exception");
+        } catch (InterruptedException e) {
+            Log.e(LOGTAG, "OpenCamera - Interrupted Exception");
+        }
+        return true;
+    }
+
+    private final CameraDevice.StateCallback mStateCallback = new CameraDevice.StateCallback() {
+
+        @Override
+        public void onOpened(CameraDevice cameraDevice) {
+            mCameraDevice = cameraDevice;
+            mCameraOpenCloseLock.release();
+            createCameraPreviewSession();
+        }
+
+        @Override
+        public void onDisconnected(CameraDevice cameraDevice) {
+            cameraDevice.close();
+            mCameraDevice = null;
+            mCameraOpenCloseLock.release();
+        }
+
+        @Override
+        public void onError(CameraDevice cameraDevice, int error) {
+            cameraDevice.close();
+            mCameraDevice = null;
+            mCameraOpenCloseLock.release();
+        }
+
+    };
+
+    private void createCameraPreviewSession() {
+        final int w=mPreviewSize.getWidth(), h=mPreviewSize.getHeight();
+        Log.i(LOGTAG, "createCameraPreviewSession("+w+"x"+h+")");
+        if(w<0 || h<0)
+            return;
+        try {
+            mCameraOpenCloseLock.acquire();
+            if (null == mCameraDevice) {
+                mCameraOpenCloseLock.release();
+                Log.e(LOGTAG, "createCameraPreviewSession: camera isn't opened");
+                return;
+            }
+            if (null != mCaptureSession) {
+                mCameraOpenCloseLock.release();
+                Log.e(LOGTAG, "createCameraPreviewSession: mCaptureSession is already started");
+                return;
+            }
+
+            mImageReader = ImageReader.newInstance(w,h,ImageFormat.YUV_420_888,2);
+            mImageReader.setOnImageAvailableListener(new ImageReader.OnImageAvailableListener() {
+                @Override
+                public void onImageAvailable(ImageReader reader) {
+                    Image image = reader.acquireLatestImage();
+                    if (image == null) return;
+                    ByteBuffer plane = image.getPlanes()[0].getBuffer();
+                    Mat yuv = new Mat( h * 3/2, w, CvType.CV_8UC1, plane );
+                    JavaCamera2Frame tempFrame = new JavaCamera2Frame(yuv,w,h);
```

Comment:
[This](https://developer.android.com/reference/android/media/Image.html#getFormat()) says that all three planes should be used for `YUV_420_888`.

BTW, Please add additional sanity checks (assertions) - for actual `Image` format, `ByteBuffer` size, etc.

Reply:
Hmm... OTOH, you're correct (there are 3 planes), but obviously, the planes are contiguous in memory and also mapped in one single chunk, otherwise, the `cvtColor` call would cause a segfault. 

To implement this in a way that references all 3 planes separately would AFAICT require a new implementation of `cvtColor` that can handle 3 separate `Mat` objects. Otherwise, you'd need to copy the planes together into a single `Mat` and lose the advantage of the zero-copy native buffer.

Reply:
We have `hal::` conversion functions accepting two separate pointers for Y and UV planes ([example](https://github.com/opencv/opencv/blob/ba940dec38983cb35de6342d5e770eb0f054549e/modules/imgproc/include/opencv2/imgproc/hal/hal.hpp#L195-L199), so you can try to use them. You can add more overloaded functions if needed. But it is hard to say how it will work taking in account that each plane has its own `PixelStride` parameter which is not supported by current implementations.

Reply:
According to https://developer.android.com/reference/android/graphics/ImageFormat.html#YUV_420_888 the Y plane will always have a stride of 1, and the U/V planes will have equal stride. I may simply assert that U/V stride is also 1 for the moment (seems to be true for all implementations I've seen so far).

Reply:
Are you sure U and V have PixelStride=1? Probably it is 2. If your current code has worked with YUV_BGRA_NV12 conversion by providing only _Y_ pointer, then planes should've been located in continuous memory block like this:
```
YYYY
YYYY
UVUV
```
Then: 
```
Uptr = Yptr + row_stride * rows
Upixel_stride = 2
Vptr = Uptr + 1
Vpixel_stride = 2
```

Reply:
Corrections to my previous comment: the conversion you used is YUV2RGB2_NV21 and in memory layout the V channel stands before U.

Reply:
But I'm using COLOR_YUV2RGB_YV12 (see below), and that is for 3 sequential planes, isn't it?

Reply:
As I can see from the code:
```.cpp
private int mPreviewFormat = ImageFormat.NV21;
```
This field is not modified anywhere and is used to determine conversion mode.

Reply:
My mistake, you're right. So it's one separate Y plane and one interleaved U/V plane.

Reply:
But in that case, `cvtTwoPlaneYUVtoBGR` should probably work?

Reply:
Probably it should, but we should check these assumptions (via `CV_Assert`) before calling any processing code.

Reply:
@floe , the problem with `cvtTwoPlaneYUVtoBGR` is that it does not get wrapped automatically, because it accepts raw C++ pointers. So you will need to write JNI wrapper by hand or wait_for/propose extended `cvtColor(InputArray ...)` interface.

Reply:
Is there any way to check whether the planes are contiguous in memory before passing them to `Mat()`? Then the current approach would already be sufficient...